lib/buffer-writer.js

```diff
@@ -0,0 +1,287 @@
+ * @param {Block} block
+ */
+export const addBlock = (writer, { cid, bytes }) => {
+  const byteLength = cid.bytes.byteLength + bytes.byteLength
+  const size = varint.encode(byteLength)
+  if (writer.byteOffset + size.length + byteLength > writer.bytes.byteLength) {
+    throw new RangeError('Buffer has no capacity for this block')
+  } else {
+    writeBytes(writer, size)
+    writeBytes(writer, cid.bytes)
+    writeBytes(writer, bytes)
+  }
+}
+
+/**
+ * @param {CarBufferWriter} writer
+ * @param {object} [options]
+ * @param {boolean} [options.resize]
+ */
+export const close = (writer, { resize = false } = {}) => {
+  const { roots, bytes, byteOffset, headerSize } = writer
+
+  const headerBytes = CBOR.encode({ version: 1, roots })
+  const varintBytes = varint.encode(headerBytes.length)
+
+  const size = varintBytes.length + headerBytes.byteLength
+  const offset = headerSize - size
+
+  // If header size estimate was accurate we just write header and return
+  // view into buffer.
+  if (offset === 0) {
+    writeHeader(writer, varintBytes, headerBytes)
+    return bytes.subarray(0, byteOffset)
+    // If header was overestimated and `{resize: true}` is passed resize header
+  } else if (resize) {
+    resizeHeader(writer, size)
+    writeHeader(writer, varintBytes, headerBytes)
+    return bytes.subarray(0, writer.byteOffset)
+  } else {
+    throw new RangeError(`Header size was overestimated.
+You can use close({ resize: true }) to resize header`)
+  }
+}
+
+/**
+ *
+ * @param {CarBufferWriter} writer
+ * @param {number} byteLength
+ */
+export const resizeHeader = (writer, byteLength) => {
+  const { bytes, headerSize } = writer
+  // Move data section to a new offset
+  bytes.set(bytes.subarray(headerSize, writer.byteOffset), byteLength)
+  // Update header size & byteOffset
+  writer.byteOffset += byteLength - headerSize
+  writer.headerSize = byteLength
+}
+
+/**
+ *
+ * @param {CarBufferWriter} writer
+ * @param {number[]|Uint8Array} bytes
+ */
+
+const writeBytes = (writer, bytes) => {
+  writer.bytes.set(bytes, writer.byteOffset)
+  writer.byteOffset += bytes.length
+}
+/**
+ *
+ * @param {{bytes:Uint8Array}} writer
+ * @param {number[]} varint
+ * @param {Uint8Array} header
+ */
+const writeHeader = ({ bytes }, varint, header) => {
+  bytes.set(varint)
+  bytes.set(header, varint.length)
+}
+
+/**
+ * Attempts to estimate header size given number of roots it will have assuming
+ * they're CIDv1 in with sha256 digest. Optionally it takes number of bytes
+ * to be allocated for all the roots, in case different hashing or CID version
+ * is used.
+ *
+ * Note: Returned value is just an estimate which can be inaccurate where large
+ * number of CIDs is passed or if they are of various sizes.
+ *
+ * @param {number} count - Number of roots
+ * @param {number} [rootsByteLength] - Total byteLength of all roots
+ */
+export const estimateHeaderSize = (
+  count,
+  rootsByteLength = count * DEFAULT_CID_SIZE
```

Comment:
This argument is still bothering me because it's getting in the way of a more accurate calculation when we can do it.

To illustrate .. and get us closer to actually doing it .. I've added a calculator in https://github.com/rvagg/cborg/pull/48 and then adapted the most efficient form of it against this branch in https://github.com/ipld/js-car/pull/71 (failing because it's not merged and compiled yet but it does pass if wired up).

But you can see from that, that we need individual root lengths, not just an aggregate. An aggregate might be good for this case where we're assuming a standard CID length, but it fails for novelty outside of that.

How would this PR change if we merged #71 (or some variation of it) into here and had an accurate calculator. What API would you present to the user with an accurate calculator but not requiring they provide the CIDs up front?

My preference would be to strip away all the estimation, present the user with an accurate calculator based on lengths and put the onus on them to provide us with either an accurate length, or an estimation that is at least long enough—but it's their responsibility to deal with that.

Reply:
> But you can see from that, that we need individual root lengths, not just an aggregate. An aggregate might be good for this case where we're assuming a standard CID length, but it fails for novelty outside of that

It is incorrect, however sometimes you don't know how what the roots will look like. So this is meant to just provide estimate. If we'd known the roots ahead we would not need to estimate anything. I also can't imagine case where I'd know CID sizes ahead but not what they are.

Reply:
> My preference would be to strip away all the estimation, present the user with an accurate calculator based on lengths and put the onus on them to provide us with either an accurate length, or an estimation that is at least long enough—but it's their responsibility to deal with that.

If you want to go with estimator as in #71 that's ok by me as well. In common case you'd just `new Array(count).fill(avgSize)` I suppose.

Reply:
I could change `estimateHeaderSize` to take `rootLengths` instead even without changes to cborg. I'm also not sure how all those allocations used by estimator will impact performance.

Reply:
I think at the end of the day I think `estimateHeaderSize` is not a great API if you do know your root sizes, if you're just guessing I'm not sure `calculateHeaderLength` will do any better.

However since one could do:

```js
const estimateHeaderSize = (count, rootsByteLength) =>
  calculateHeaderLength(new Array(count).fill(rootsByteLength / count))
```

But it's not possible to do other way round, I'm convinced that replacing `estimateHeaderSize` with `calculateHeaderLength` is good idea.

Reply:
> I also can't imagine case where I'd know CID sizes ahead but not what they are

Yeah, but my main problem here is with the second argument and it assumes you do know the sizes. It's foot-gunny and invites bad assumptions. Maybe better to just have the one argument and say that the estimator assumes typical CID sizes, i.e. hide the second argument inside the function, not as an optional argument with a default. _Or_ make the second argument a singular CID length that gets multiplied by the first—by doing that you could account for CIDv0's and CIDs with hashers with multicodecs in the multi-byte range (like Filecoin's blake2b256).

> I'm also not sure how all those allocations used by estimator will impact performance

my concern too, but mocha helps out a bit with measuring that:

```
      ✔ estimateHeaderCapacity(65535) (98ms)
      ✔ calculateHeaderLength(65535) (84ms)
      ✔ estimateHeaderCapacity(65536) (62ms)
      ✔ calculateHeaderLength(65536) (79ms)
```

(very variable timings, they flip around quite a bit, this is a random grab that happens to look good)

(a) the VM is doing a decent job of optimising away some of the allocations (`new` really isn't that scary) and (b) there really aren't that many allocations going on and the call depth involved with the calculator is actually pretty simple, just hidden by abstractions to account for more complex cases, so a lot of that gets optimised away too.

---

What about both? `estimateHeaderSize(count, [cidLength])` for the case where you just want a simple value and know the number of roots and _maybe_ have some novel characteristics about your CIDs, and `calculateHeaderSize(rootLengths)` for the case where you have some certainty.